File: code/1466.reorder-routes-to-make-all-paths-lead-to-the-city-zero.php
<?php
/*
 * @lc app=leetcode id=1466 lang=php
 *
 * [1466] Reorder Routes to Make All Paths Lead to the City Zero
 */

// @lc code=start
class Solution {
    private $count = 0;
    /**
     * @param Integer $n
     * @param Integer[][] $connections
     * @return Integer
     */
    function minReorder($n, $connections) {
        $visited = [];
        $graph = [];
        for ($i = 0; $i < $n; $i++) {
            $graph[] = [];
        }
        foreach ($connections as $connection) {
            $graph[$connection[0]][] = $connection[1];
            $graph[$connection[1]][] = -$connection[0];
        }

        $this->helper($graph, $visited, 0);
        return $this->count;
    }

    function helper($graph, &$visited, $from) {
        $visited[$from] = true;
        foreach ($graph[$from] as $to) {
            if (!$visited[abs($to)]) {
                if ($to > 0) $this->count++;
                $this->helper($graph, $visited, abs($to));
            }
        }
    }
}
// @lc code=end

File: code/1466.reorder-routes-to-make-all-paths-lead-to-the-city-zero.go
package code

/*
 * @lc app=leetcode id=1466 lang=golang
 *
 * [1466] Reorder Routes to Make All Paths Lead to the City Zero
 */

// @lc code=start
func minReorder(n int, connections [][]int) int {
	visited := make([]bool, n)

	graph := make([][]int, n)
	for _, connection := range connections {
		graph[connection[0]] = append(graph[connection[0]], connection[1])
		graph[connection[1]] = append(graph[connection[1]], -connection[0])
	}

	return rec1466(graph, visited, 0)
}

func rec1466(graph [][]int, visited []bool, from int) int {
	visited[from] = true
	reorderCount := 0
	for _, to := range graph[from] {
		if visited[abs(to)] {
			continue
		}
		if to > 0 {
			reorderCount++
		}
		reorderCount += rec1466(graph, visited, abs(to))
	}
	return reorderCount
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// @lc code=end
